pkg/kkrtpsi/sender.go

```diff
@@ -0,0 +1,211 @@
+}
+
+// oprfEncodedInputs stores the possible bucket
+// indexes in the receiver cuckoo hash table
+type oprfEncodedInputs struct {
+	prcEncoded [cuckoo.Nhash][]byte // PseudoRandom Code
+	bucketIdx  [cuckoo.Nhash]uint64
+}
+
+// inputsAndHasher is used to pass the OPRF encoded
+// inputs along with the hasher from stage 1 to stage 3
+type inputsAndHasher struct {
+	inputs []oprfEncodedInputs
+	hasher hash.Hasher
+}
+
+// NewSender returns a KKRTPSI sender initialized to
+// use rw as the communication layer
+func NewSender(rw io.ReadWriter) *Sender {
+	return &Sender{rw: rw}
+}
+
+// Send initiates a KKRTPSI exchange
+// that reads local IDs from identifiers, until identifiers closes.
+// The format of an indentifier is string
+// example:
+//  0e1f461bbefa6e07cc2ef06b9ee1ed25101e24d4345af266ed2f5a58bcd26c5e
+func (s *Sender) Send(ctx context.Context, n int64, identifiers <-chan []byte) (err error) {
+	// fetch and set up logger
+	logger := logr.FromContextOrDiscard(ctx)
+	logger = logger.WithValues("protocol", "kkrtpsi")
+
+	// statistics
+	start := time.Now()
+	timer := time.Now()
+	var mem uint64
+
+	var seeds [cuckoo.Nhash][]byte
+	var remoteN int64     // receiver size
+	var oprfInputSize int // nb of OPRF keys
+
+	var oprfKeys *oprf.Key
+	var encodedInputChan = make(chan inputsAndHasher)
+
+	// stage 1: sample 3 hash seeds and write them to receiver
+	// for cuckoo hashing parameters agreement.
+	// read local ids and store the potential bucket indexes for each id.
+	stage1 := func() error {
+		logger.V(1).Info("Starting stage 1")
+
+		// sample Nhash hash seeds
+		for i := range seeds {
+			seeds[i] = make([]byte, hash.SaltLength)
+			if _, err := rand.Read(seeds[i]); err != nil {
+				return err
+			}
+			// write it into rw
+			if _, err := s.rw.Write(seeds[i]); err != nil {
+				return err
+			}
+		}
+
+		// read remote input size
+		if err := binary.Read(s.rw, binary.BigEndian, &remoteN); err != nil {
+			return err
+		}
+
+		// sample random 16 byte secret key for AES-128 and send to the receiver
+		sk := make([]byte, aes.BlockSize)
+		if _, err = rand.Read(sk); err != nil {
+			return err
+		}
+
+		// send the secret key
+		if _, err := s.rw.Write(sk); err != nil {
+			return err
+		}
+
+		// calculate number of OPRF from the receiver based on
+		// number of buckets in cuckooHashTable
+		oprfInputSize = int(cuckoo.Factor * float64(remoteN))
+		if 1 > oprfInputSize {
+			oprfInputSize = 1
+		}
+
+		// exhaust local ids, and precompute all potential
+		// hashes and store them using the same
+		// cuckoo hash table parameters as the receiver.
+		go func() {
+			cuckooHasher, err := cuckoo.NewCuckooHasher(uint64(remoteN), seeds)
+			if err != nil {
+				panic(err)
+			}
+			// instantiate an AES block
+			aesBlock, err := aes.NewCipher(sk)
+			if err != nil {
+				panic(err)
+			}
+
+			// prepare struct to send inputs and hasher to stage 3
+			var message inputsAndHasher
+			message.inputs = make([]oprfEncodedInputs, n)
+			i := 0
```

Comment:
You can use 2 variables in range to get the index as well.

Reply:
Actually in Go the index variable is not allowed when ranging over a channel. That is why I have a separate index variable here.

Reply:
Actually Justin had a good idea to avoid ranging over the channel entirely